okhttp/src/jvmTest/java/okhttp3/internal/publicsuffix/PublicSuffixListGenerator.kt

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.publicsuffix
+
+import java.util.SortedSet
+import java.util.TreeSet
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.executeAsync
+import okhttp3.internal.publicsuffix.PublicSuffixDatabase.Companion.PUBLIC_SUFFIX_RESOURCE
+import okio.BufferedSink
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+import okio.FileSystem
+import okio.Path
+import okio.Path.Companion.toPath
+import okio.buffer
+import okio.gzip
+
+/**
+ * Downloads the public suffix list from https://publicsuffix.org/list/public_suffix_list.dat and
+ * transforms the file into an efficient format used by OkHttp.
+ *
+ *
+ * The intent is to use this class to update the list periodically by manually running the main
+ * method. This should be run from the top-level okhttp directory.
+ *
+ *
+ * The resulting file is used by [PublicSuffixDatabase].
+ */
+class PublicSuffixListGenerator(
+  projectRoot: Path = ".".toPath(),
+  val fileSystem: FileSystem = FileSystem.SYSTEM,
+  val client: OkHttpClient = OkHttpClient()
+) {
+  private val resources = projectRoot / "okhttp/src/jvmMain/resources/okhttp3/internal/publicsuffix"
+  private val testResources = projectRoot / "okhttp/src/jvmTest/resources/okhttp3/internal/publicsuffix"
+  private val publicSuffixListDotDat = testResources / "public_suffix_list.dat"
+  private val outputFile = resources / PUBLIC_SUFFIX_RESOURCE
+
+  val request = Request("https://publicsuffix.org/list/public_suffix_list.dat".toHttpUrl())
+
+  suspend fun import() {
+    check(fileSystem.metadata(resources).isDirectory)
+    check(fileSystem.metadata(testResources).isDirectory)
+
+    updateLocalFile()
```

Comment:
Having each of these functions do `withContext(Dispatchers.IO)`, rather than pulling that dispatcher switch up here is surprising? As-is it’s correct and good though.

Reply:
I think it's that thing about making all suspend functions main safe, and that `withContext(Dispatchers.IO)` should be a no-op from Dispatchers.IO or Dispatchers.Default.